{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read CSV for diNT PWM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "diNT_csv_dictionary = {}\n",
    "\n",
    "#opens csv file, reads into a dictionary\n",
    "with open('diNTpwm_fordict.csv') as csvfile:\n",
    "    reader = csv.DictReader(csvfile)\n",
    "    for row in reader:\n",
    "        diNT_csv_dictionary[row['dinucleotide'], row['position']] = row['efficiency']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# diNT PWM score TIS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#1 ,  2,  3,  4,  5,  6, 7\n",
    "#-6, -5, -4, -3, -2, -1, +4, +5\n",
    "\n",
    "#have to scan through each dinucleotide position in seq\n",
    "#get the efficiency and multiply together\n",
    "#save into table\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO\n",
    "\n",
    "\n",
    "# 30_ATG_rest\n",
    "\n",
    "TIS_list = []\n",
    "#get the sequence out of the fasta file # should start at 24 until 35\n",
    "for seq_record in SeqIO.parse(\"finalviral.fasta\", \"fasta\"):\n",
    "    #transcribes from DNA to RNA by replacing T with U\n",
    "    TIS_list.append(seq_record.seq[24:35].transcribe())\n",
    "  \n",
    "    \n",
    "    \n",
    "\n",
    "TIS_list = []\n",
    "#get the sequence out of the fasta file # should start at 24 until 35\n",
    "for seq_record in SeqIO.parse(\"hs_all_nodup_u36.fasta\", \"fasta\"):\n",
    "    #transcribes from DNA to RNA by replacing T with U\n",
    "    TIS_list.append(seq_record.seq.upper()[30:41].transcribe())\n",
    "\n",
    "    \n",
    "    \n",
    "\n",
    "def PWM(dinucleotide, position):\n",
    "    #convert efficiency into float\n",
    "    return float(diNT_csv_dictionary[dinucleotide, position])\n",
    "\n",
    "seq_dint_tuple_list = []\n",
    "for i in range(len(TIS_list)):\n",
    "\n",
    "    for j in range(0,5):\n",
    "        seq_dint_tuple_list.append((str(TIS_list[i][j:j+2]), j+1))\n",
    "\n",
    "    # add dint position 6\n",
    "    seq_dint_tuple_list.append((TIS_list[i][5]+TIS_list[i][9], 6))\n",
    "\n",
    "\n",
    "    # add dint position 7\n",
    "    seq_dint_tuple_list.append((TIS_list[i][9]+TIS_list[i][10], 7))\n",
    "    \n",
    "\n",
    "\n",
    "#first index is 0 to 6\n",
    "#2nd index is from 0 to 1\n",
    "\n",
    "#have to accumulate every 7 PWMs\n",
    "\n",
    "PWM_diNT_scores_seq = []\n",
    "for i in range(len(seq_dint_tuple_list)):\n",
    "    PWM_diNT_scores_seq.append(PWM(seq_dint_tuple_list[i][0], str(seq_dint_tuple_list[i][1])))\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    \n",
    "list_of_tuples_by_seven = [PWM_diNT_scores_seq[x:x+7] for x in range(0, len(PWM_diNT_scores_seq),7)]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# # have to multiply every 7 indices\n",
    "    \n",
    "import operator\n",
    "import functools\n",
    "\n",
    "PWM_scores_seq = []\n",
    "for i in range(len(list_of_tuples_by_seven)):\n",
    "    PWM_scores_seq.append(functools.reduce(operator.mul, list_of_tuples_by_seven[i], 7))\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "#convert to pandas series to easily manipulate\n",
    "pandas_PWM_scores_human = pd.Series(PWM_scores_seq)\n",
    "\n",
    "import math\n",
    "final_data_human = pandas_PWM_scores_human*math.exp(4.602730)\n",
    "\n",
    "plt.hist(final_data_human, bins = 200)\n",
    "plt.show()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#convert to pandas series to easily manipulate\n",
    "pandas_PWM_scores_viral = pd.Series(PWM_scores_seq)\n",
    "\n",
    "import math\n",
    "final_data_viral = pandas_PWM_scores_viral*math.exp(4.602730)\n",
    "\n",
    "plt.hist(final_data_viral, bins = 200)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Randomize TIS sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "#for human sequences\n",
    "from Bio import SeqIO\n",
    "TIS_list = []\n",
    "#get the sequence out of the fasta file # should start at 24 until 35\n",
    "for seq_record in SeqIO.parse(\"hs_all_nodup_u36.fasta\", \"fasta\"):\n",
    "    #transcribes from DNA to RNA by replacing T with U\n",
    "    TIS_list.append(seq_record.seq.upper()[30:41].transcribe())\n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "#rerun to get more random seq\n",
    "import numpy\n",
    "import random\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_rna\n",
    "\n",
    "shuffled_TIS_list =[]\n",
    "\n",
    "for each in TIS_list:\n",
    "    TIS_only = str(each[0:6] + each[9:11])\n",
    "    \n",
    "    #shuffles a string\n",
    "    shuffled_TIS_source_string = ''.join(random.sample(TIS_only,len(TIS_only)))\n",
    "\n",
    "    \n",
    "    shuffled_record = Seq(shuffled_TIS_source_string[0:6] + \"AUG\" + shuffled_TIS_source_string[6:8], \n",
    "            generic_rna)\n",
    "    shuffled_TIS_list.append(shuffled_record)\n",
    "\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "seq_dint_tuple_list = []\n",
    "for i in range(len(shuffled_TIS_list)):\n",
    "\n",
    "    for j in range(0,5):\n",
    "        seq_dint_tuple_list.append((str(shuffled_TIS_list[i][j:j+2]), j+1))\n",
    "\n",
    "    # add dint position 6\n",
    "    seq_dint_tuple_list.append((shuffled_TIS_list[i][5]+TIS_list[i][9], 6))\n",
    "\n",
    "\n",
    "    # add dint position 7\n",
    "    seq_dint_tuple_list.append((shuffled_TIS_list[i][9]+TIS_list[i][10], 7))\n",
    "    \n",
    "\n",
    "    \n",
    "    #first index is 0 to 6\n",
    "#2nd index is from 0 to 1\n",
    "\n",
    "#have to accumulate every 7 PWMs\n",
    "\n",
    "PWM_diNT_scores_seq = []\n",
    "for i in range(len(seq_dint_tuple_list)):\n",
    "    PWM_diNT_scores_seq.append(PWM(seq_dint_tuple_list[i][0], str(seq_dint_tuple_list[i][1])))\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    \n",
    "list_of_tuples_by_seven = [PWM_diNT_scores_seq[x:x+7] for x in range(0, len(PWM_diNT_scores_seq),7)]\n",
    "\n",
    "\n",
    "# # have to multiply every 7 indices\n",
    "    \n",
    "import operator\n",
    "import functools\n",
    "\n",
    "PWM_scores_seq = []\n",
    "for i in range(len(list_of_tuples_by_seven)):\n",
    "    PWM_scores_seq.append(functools.reduce(operator.mul, list_of_tuples_by_seven[i], 7))\n",
    "\n",
    "\n",
    "    \n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "#convert to pandas series to easily manipulate\n",
    "pandas_PWM_scores_human = pd.Series(PWM_scores_seq)\n",
    "\n",
    "import math\n",
    "shuffled_human = pandas_PWM_scores_human*math.exp(4.602730)\n",
    "\n",
    "plt.hist(shuffled_human, bins = 200)\n",
    "plt.show()\n",
    "#looks the same\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot density of diNT PWM scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#plots densities of the efficiencies\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy.stats import gaussian_kde\n",
    "\n",
    "\n",
    "\n",
    "data_human = final_data_human\n",
    "\n",
    "density_human = gaussian_kde(data_human)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "data_viral = final_data_viral\n",
    "\n",
    "density_viral = gaussian_kde(data_viral)\n",
    "\n",
    "\n",
    "xs = np.linspace(-1000,4000,2000)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "density_viral.covariance_factor = lambda : .25\n",
    "\n",
    "density_viral._compute_covariance()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "density_human.covariance_factor = lambda : .25\n",
    "\n",
    "density_human._compute_covariance()\n",
    "\n",
    "\n",
    "plt.plot(xs,density_viral(xs), label = \"viral TIS\")\n",
    "plt.plot(xs,density_human(xs), label = \"human TIS\")\n",
    "plt.legend()\n",
    "plt.show()\n",
    "\n",
    "#human"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
