{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "#viral TIS | human seq | gemini virus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# TESTing old plot density\n",
    "\n",
    "# #plots densities of the efficiencies\n",
    "\n",
    "# import matplotlib.pyplot as plt\n",
    "# import numpy as np\n",
    "# from scipy.stats import gaussian_kde\n",
    "\n",
    "# xs = np.linspace(-1000,4000,2000)\n",
    "\n",
    "\n",
    "# # plt.hist(viral_data,bins = 200)\n",
    "# # plt.hist(human_data, bins = 200)\n",
    "# # plt.hist(scrambled_human_data,bins=200)\n",
    "\n",
    "\n",
    "# # density_viral = gaussian_kde(viral_data)\n",
    "# # density_viral.covariance_factor = lambda : .25\n",
    "# # density_viral._compute_covariance()\n",
    "# # plt.plot(xs,density_viral(xs), label = \"viral TIS\")\n",
    "\n",
    "\n",
    "\n",
    "# density_human = gaussian_kde(human_data)\n",
    "# density_human.covariance_factor = lambda : .25\n",
    "# density_human._compute_covariance()\n",
    "# plt.plot(xs,density_human(xs), label = \"human TIS\")\n",
    "\n",
    "\n",
    "\n",
    "# # density_scrambled_human = gaussian_kde(scram_human_data)\n",
    "# # density_scrambled_human.covariance_factor = lambda : .25\n",
    "# # density_scrambled_human._compute_covariance()\n",
    "# # plt.plot(xs,density_scrambled_human(xs),label = \"scrambled human TIS\")\n",
    "\n",
    "# # density_human_IA = gaussian_kde(HUMAN_IA_scores)\n",
    "# # density_human_IA.covariance_factor = lambda : .25\n",
    "# # density_human_IA._compute_covariance()\n",
    "# # plt.plot(xs,density_human_IA(xs),label = \"human TIS downstream AUG\")\n",
    "\n",
    "# # density_gemini = gaussian_kde(gemini_data)\n",
    "# # density_gemini.covariance_factor = lambda : .25\n",
    "# # density_gemini._compute_covariance()\n",
    "# # plt.plot(xs,density_gemini(xs),label = \"gemini TIS\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# # scanning_viral_data\n",
    "# # ires_viral_data\n",
    "# # leaky_viral_data\n",
    "# # nsp_viral_data\n",
    "# # rs_viral_data\n",
    "# # ti_viral_data\n",
    "# # vpg_viral_data\n",
    "# # unknown_viral_data\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# # density_s = gaussian_kde(scanning_viral_data)\n",
    "# # density_s.covariance_factor = lambda : .25\n",
    "# # density_s._compute_covariance()\n",
    "# # plt.plot(xs,density_s(xs),label = \"scanning viral TIS\")\n",
    "\n",
    "\n",
    "\n",
    "# # density_s_scram = gaussian_kde(scram_SCANNING_data)\n",
    "# # density_s_scram.covariance_factor = lambda : .25\n",
    "# # density_s_scram._compute_covariance()\n",
    "# # plt.plot(xs,density_s_scram(xs),label = \"scram scanning viral TIS\")\n",
    "\n",
    "\n",
    "\n",
    "# # density_i = gaussian_kde(ires_viral_data)\n",
    "# # density_i.covariance_factor = lambda : .25\n",
    "# # density_i._compute_covariance()\n",
    "# # plt.plot(xs,density_i(xs),label = \"ires viral TIS\")\n",
    "\n",
    "\n",
    "# # density_i_scram = gaussian_kde(scram_IRES_data)\n",
    "# # density_i_scram.covariance_factor = lambda : .25\n",
    "# # density_i_scram._compute_covariance()\n",
    "# # plt.plot(xs,density_i_scram(xs),label = \"scram ires viral TIS\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# # density_l = gaussian_kde(leaky_viral_data)\n",
    "# # density_l.covariance_factor = lambda : .25\n",
    "# # density_l._compute_covariance()\n",
    "# # plt.plot(xs,density_l(xs),label = \"leaky viral TIS\")\n",
    "\n",
    "\n",
    "# # density_l_scram = gaussian_kde(scram_LEAKY_data)\n",
    "# # density_l_scram.covariance_factor = lambda : .25\n",
    "# # density_l_scram._compute_covariance()\n",
    "# # plt.plot(xs,density_l_scram(xs),label = \"scram leaky viral TIS\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# # density_n = gaussian_kde(nsp_viral_data)\n",
    "# # density_n.covariance_factor = lambda : .25\n",
    "# # density_n._compute_covariance()\n",
    "# # plt.plot(xs,density_n(xs),label = \"nsp viral TIS\")\n",
    "\n",
    "\n",
    "\n",
    "# # density_n_scram = gaussian_kde(scram_NSP_data)\n",
    "# # density_n_scram.covariance_factor = lambda : .25\n",
    "# # density_n_scram._compute_covariance()\n",
    "# # plt.plot(xs,density_n_scram(xs),label = \"scram nsp viral TIS\")\n",
    "\n",
    "\n",
    "\n",
    "# # density_rs = gaussian_kde(rs_viral_data)\n",
    "# # density_rs.covariance_factor = lambda : .25\n",
    "# # density_rs._compute_covariance()\n",
    "# # plt.plot(xs,density_rs(xs),label = \"ribosome shunting viral TIS\")\n",
    "\n",
    "# # density_rs_scram = gaussian_kde(scram_RS_data)\n",
    "# # density_rs_scram.covariance_factor = lambda : .25\n",
    "# # density_rs_scram._compute_covariance()\n",
    "# # plt.plot(xs,density_rs_scram(xs),label = \"scram ribosome shunting viral TIS\")\n",
    "\n",
    "\n",
    "\n",
    "# # density_ti = gaussian_kde(ti_viral_data)\n",
    "# # density_ti.covariance_factor = lambda : .25\n",
    "# # density_ti._compute_covariance()\n",
    "# # plt.plot(xs,density_ti(xs),label = \"TR viral TIS\")\n",
    "\n",
    "# # density_ti_scram = gaussian_kde(scram_TI_data)\n",
    "# # density_ti_scram.covariance_factor = lambda : .25\n",
    "# # density_ti_scram._compute_covariance()\n",
    "# # plt.plot(xs,density_ti_scram(xs),label = \"scram TR viral TIS\")\n",
    "\n",
    "# # density_v = gaussian_kde(vpg_viral_data)\n",
    "# # density_v.covariance_factor = lambda : .25\n",
    "# # density_v._compute_covariance()\n",
    "# # plt.plot(xs,density_v(xs),label = \"vpg viral TIS\")\n",
    "\n",
    "\n",
    "# # density_v_scram = gaussian_kde(scram_VPG_data)\n",
    "# # density_v_scram.covariance_factor = lambda : .25\n",
    "# # density_v_scram._compute_covariance()\n",
    "# # plt.plot(xs,density_v_scram(xs),label = \"scram vpg viral TIS\")\n",
    "\n",
    "\n",
    "# # density_uk = gaussian_kde(unknown_viral_data)\n",
    "# # density_uk.covariance_factor = lambda : .25\n",
    "# # density_uk._compute_covariance()\n",
    "# # plt.plot(xs,density_uk(xs),label = \"unknown TIS\")\n",
    "\n",
    "\n",
    "\n",
    "# # density_uk_scram = gaussian_kde(scram_UNKNOWN_data)\n",
    "# # density_uk_scram.covariance_factor = lambda : .25\n",
    "# # density_uk_scram._compute_covariance()\n",
    "# # plt.plot(xs,density_uk_scram(xs),label = \"scram unknown TIS\")\n",
    "\n",
    "\n",
    "# density_all_possible = gaussian_kde(all_possible_data)\n",
    "# density_all_possible.covariance_factor = lambda : .25\n",
    "# density_all_possible._compute_covariance()\n",
    "# plt.plot(xs,density_all_possible(xs),label = \"all possible TIS\")\n",
    "\n",
    "\n",
    "# plt.title(\"Scanning vs human\")\n",
    "# plt.legend()\n",
    "# plt.show()\n",
    "\n",
    "# #human"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# TODO\n",
    "\n",
    "\n",
    "\n",
    "# compare human and shuffled seq\n",
    "# scramble viral seq\n",
    "# scrambled\n",
    "\n",
    "\n",
    "# 4 distributions\n",
    "\n",
    "# plant viruses\n",
    "\n",
    "\n",
    "# human human viruses | arabidopsis and (tomato) arabidopsis viruses\n",
    "\n",
    "\n",
    "# methods of RBSdesigner\n",
    "\n",
    "# human vs viral for all variants\n",
    "# will have many distribustion\n",
    "\n",
    "# human vs human shuffle\n",
    "# human vs viral shuffle\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "# from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "# from Bio import pairwise2\n",
    "# import warnings\n",
    "# from Bio import BiopythonWarning\n",
    "# import re\n",
    "import csv\n",
    "\n",
    "\n",
    "\n",
    "# make a dictionary of the csv file and clean up the accession numbers so it matches the viral data\n",
    "#----------------------------------------------------------------------------------------------------------------\n",
    "reader = csv.DictReader(open('TIS_annotation.csv'))\n",
    "\n",
    "TIS_annotation_dict = {}\n",
    "\n",
    "for row in reader:\n",
    "    \n",
    "    #get old key\n",
    "    key  = row['Accession_#']\n",
    "    start = key.find('NC')\n",
    "    end = key.find('.', start)\n",
    "    \n",
    "    #strip out the verision number\n",
    "    cleaned_key = key[start:end]\n",
    "    \n",
    "    #makes a new correct TIS annotation dictionary\n",
    "    TIS_annotation_dict[cleaned_key] = row['primary_TIS_type']\n",
    "#----------------------------------------------------------------------------------------------------------------\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2528"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "# annotate TIS with TIS type\n",
    "#----------------------------------------------------------------------------------------------------------------\n",
    "annotated_extracted_TIS_list = []\n",
    "\n",
    "filename = \"fixed_accession_viral_all_30up_toSTOP.fasta\"\n",
    "\n",
    "#for every record in the fasta file, delimited by >\n",
    "for record in SeqIO.parse(filename, \"fasta\"):\n",
    "    \n",
    "    record_description = str(record.description)\n",
    "    \n",
    "    #make a new record with the sequence, name of protein, and the TIS type\n",
    "    annotated_record = SeqRecord(record.seq,\n",
    "                                 record.name,\n",
    "                                 description = record_description + str(TIS_annotation_dict.get(record.id)))\n",
    "\n",
    "    annotated_extracted_TIS_list.append(annotated_record)\n",
    "\n",
    "#----------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# IRES\n",
    "# IRES, Leaky Scanning\n",
    "# Leaky Scanning\n",
    "# NSP3-3'UTR\n",
    "# Ribosome Shunting\n",
    "# Scanning\n",
    "# Scanning, IRES\n",
    "# Termination-Reinitiation\n",
    "# VPg\n",
    "# UNKNOWN\n",
    "\n",
    "\n",
    "\n",
    "#have to separate it out into different files\n",
    "\n",
    "\n",
    "\n",
    "# write out annotated TIS list to file\n",
    "SeqIO.write(annotated_extracted_TIS_list, \"annotated_\" + filename, \"fasta\")\n",
    "\n",
    "\n",
    "\n",
    "# def set_rna_TIS(fullpath, filename):\n",
    "\n",
    "#     annotated_extracted_TIS_list = []\n",
    "\n",
    "#     for record in SeqIO.parse(fullpath, \"fasta\"):\n",
    "#         record_description = str(record.description)\n",
    "#                                                                                                 #you can change this\n",
    "#         annotated_record = SeqRecord(record.seq, record.name,description = record_description + \"Scanning\")\n",
    "\n",
    "#         annotated_extracted_TIS_list.append(annotated_record)\n",
    "        \n",
    "#     SeqIO.write(annotated_extracted_TIS_list, \"annotated_\" + filename +\".fasta\", \"fasta\")\n",
    "#     return\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# creates a list of the files in this directory\n",
    "# raw_datadir_listing = os.listdir(os.getcwd())\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# loops over the list of files\n",
    "\n",
    "\n",
    "# for files in raw_datadir_listing:\n",
    "#     if files.startswith('viral_all_30up_toSTOP'):\n",
    "#         full_path = os.path.join(os.getcwd(), files)\n",
    "#         filename = os.path.splitext(files)[0]\n",
    "#         set_viral_TIS(full_path, filename)\n",
    "        \n",
    "\n",
    "    # if files.startswith('extracted_TIS_rna'):\n",
    "    #     full_path = os.path.join(os.getcwd(), files)\n",
    "    #     filename = os.path.splitext(files)[0]\n",
    "    #     set_rna_TIS(full_path, filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# CODE FOR INTERACE WITH ENTREZ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "#search ncbi GenBank with term biopython\n",
    "handle = Entrez.esearch(db =\"nucleotide\", term = \"biopython\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['28011774', '24929426', '24497503', '24267035', '24194598', '23842806', '23157543', '22909249', '22399473', '21666252', '21210977', '20015970', '19811691', '19773334', '19304878', '18606172', '21585724', '16403221', '16377612', '14871861']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "record = Entrez.read(handle)\n",
    "record[\"IdList\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Fix unnown accession number in original fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "\n",
    "Entrez.email = \"uoafjDASLFJLASJDF9WEA@gmail.com\"\n",
    "\n",
    "\n",
    "fixed_accession_TIS_list = []\n",
    "\n",
    "\n",
    "for seq_record_local in SeqIO.parse(\"viral_all_30up_toSTOP.fasta\", \"fasta\"):\n",
    "    \n",
    "    \n",
    "    #copy all seq_record_local to new file \n",
    "    \n",
    "    \n",
    "    if \"unknown\" in seq_record_local.name:\n",
    "        \n",
    "        \n",
    "        # get the protein name per sequence\n",
    "        seq_description = seq_record_local.description\n",
    "\n",
    "        start = seq_description.find('P_')\n",
    "        end = seq_description.find('|', start)\n",
    "\n",
    "        full_protein_name = seq_description[start-1:end]\n",
    "        \n",
    "\n",
    "        handle = Entrez.efetch(db=\"protein\", rettype=\"gp\", retmode=\"text\",\n",
    "                       id=full_protein_name)\n",
    "        \n",
    "        \n",
    "        # search NCBI protein db and update all missing genome accession numbers\n",
    "        \n",
    "        for protein_id_remote in SeqIO.parse(handle, \"genbank\"):\n",
    "            \n",
    "            #get the accession number that links to the protein ID in the ncbi db\n",
    "            accession_num_unstrip = protein_id_remote.annotations[\"db_source\"]\n",
    "            \n",
    "            start = accession_num_unstrip.find('NC_')\n",
    "            end = accession_num_unstrip.find('.',start)\n",
    "            \n",
    "            accession_num = accession_num_unstrip[start:end]\n",
    "            \n",
    "            \n",
    "            seq_record_local.id = accession_num\n",
    "            seq_record_local.name = accession_num\n",
    "            seq_record_local.description = \"|\" +full_protein_name +\"|\"\n",
    "\n",
    "            \n",
    "            fixed_accession_record = SeqRecord(seq_record_local.seq,\n",
    "                                               id = seq_record_local.id,\n",
    "                                               name = seq_record_local.name, \n",
    "                                               description = seq_record_local.description)\n",
    "\n",
    "            fixed_accession_TIS_list.append(fixed_accession_record)\n",
    "            \n",
    "#             print(\"changed seq record\")\n",
    "\n",
    "        handle.close()\n",
    "    else:\n",
    "        \n",
    "        #else append the unchanged seq_record into the fixed_accession_TIS_list\n",
    "        \n",
    "        fixed_accession_TIS_list.append(seq_record_local)\n",
    "#         print(\"unchanged seq record\")\n",
    "        \n",
    "        \n",
    "#have to write the fixed unknown accession number to file\n",
    "\n",
    "\n",
    "\n",
    "SeqIO.write(fixed_accession_TIS_list, \"fixed_accession_\" + \"viral_all_30up_toSTOP.fasta\", \"fasta\")\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# len(fixed)\n",
    "len(fixed_accession_TIS_list)\n",
    "\n",
    "#should have 2528 sequences\n",
    "\n",
    "count = 0\n",
    "for seq_record in SeqIO.parse(\"annotated_fixed_accession_viral_all_30up_toSTOP.fasta\", \"fasta\"):\n",
    "    count = count+1\n",
    "    print(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_004108 |NP_671968.1|IRES or Leaky Scanning\n",
      "NC_004109 |NP_671969.1|IRES or Leaky Scanning\n",
      "NC_004110 |NP_671970.1|IRES or Leaky Scanning\n",
      "NC_004110 |NP_671971.1|IRES or Leaky Scanning\n",
      "NC_005218 |NP_941977.1|Scanning or IRES\n",
      "NC_005219 |NP_941978.1|Scanning or IRES\n",
      "NC_005222 |NP_941982.1|Scanning or IRES\n"
     ]
    }
   ],
   "source": [
    "#some fasta will contain duplicates b/c of sharing TIS types\n",
    "\n",
    "Scanning = []\n",
    "\n",
    "\n",
    "IRES = []\n",
    "\n",
    "LS = []\n",
    "NSP = []\n",
    "RS = []\n",
    "\n",
    "TR = []\n",
    "\n",
    "VPG = []\n",
    "\n",
    "UNKNOWN = []\n",
    "\n",
    "\n",
    "# how many delimiters in to split\n",
    "#2nd delimiter\n",
    "n = 2\n",
    "\n",
    "for seq_record in SeqIO.parse(\"annotated_fixed_accession_viral_all_30up_toSTOP.fasta\", \"fasta\"):\n",
    "    \n",
    "    #split string after 2nd delimiter\n",
    "    groups = seq_record.description.split('|')\n",
    "    TIS_type = '|'.join(groups[n:])\n",
    "    \n",
    "    \n",
    "    if TIS_type == \"Scanning\":\n",
    "        Scanning.append(seq_record)\n",
    "    elif TIS_type == \"Scanning, IRES\":\n",
    "        Scanning.append(seq_record)\n",
    "        IRES.append(seq_record)\n",
    "    elif TIS_type == \"IRES\":\n",
    "        IRES.append(seq_record)\n",
    "#     elif TIS_type == \"IRES, Leaky Scanning\":\n",
    "#         IRES.append(seq_record)\n",
    "#         LS.append(seq_record)\n",
    "    elif TIS_type == \"Leaky Scanning\":\n",
    "        LS.append(seq_record)\n",
    "    elif TIS_type == \"NSP3-3'UTR\":\n",
    "        NSP.append(seq_record)\n",
    "    elif TIS_type == \"Ribosome Shunting\":\n",
    "        RS.append(seq_record)\n",
    "    elif TIS_type == \"Termination-Reinitiation\":\n",
    "        TR.append(seq_record)\n",
    "    elif TIS_type == \"VPg\":\n",
    "        VPG.append(seq_record)\n",
    "    elif TIS_type == \"UNKNOWN\":\n",
    "        UNKNOWN.append(seq_record)\n",
    "    else:\n",
    "        print(seq_record.name)\n",
    "        print(seq_record.description)\n",
    "\n",
    "# Scanning\n",
    "# Scanning, IRES\n",
    "# IRES\n",
    "# IRES, Leaky Scanning\n",
    "# Leaky Scanning\n",
    "# NSP3-3'UTR\n",
    "# Ribosome Shunting\n",
    "\n",
    "# Termination-Reinitiation\n",
    "# VPg\n",
    "# UNKNOWN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "336"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# Scanning = []\n",
    "# IRES = []\n",
    "# LS = []\n",
    "# NSP = []\n",
    "# RS = []\n",
    "# TR = []\n",
    "# VPG = []\n",
    "# UNKNOWN = []\n",
    "\n",
    "SeqIO.write(Scanning, \"SCANNING_\" + \"viral_all_30up_toSTOP_SEP.fasta\", \"fasta\")\n",
    "SeqIO.write(IRES, \"IRES_\" + \"viral_all_30up_toSTOP_SEP.fasta\", \"fasta\")\n",
    "SeqIO.write(LS, \"LEAKY_\" + \"viral_all_30up_toSTOP_SEP.fasta\", \"fasta\")\n",
    "SeqIO.write(NSP, \"NSP_\" + \"viral_all_30up_toSTOP_SEP.fasta\", \"fasta\")\n",
    "SeqIO.write(RS, \"RIBOSOMESHUNTING_\" + \"viral_all_30up_toSTOP_SEP.fasta\", \"fasta\")\n",
    "SeqIO.write(TR, \"TERMINATIONREINITIATION_\" + \"viral_all_30up_toSTOP_SEP.fasta\", \"fasta\")\n",
    "SeqIO.write(VPG, \"VPG_\" + \"viral_all_30up_toSTOP_SEP.fasta\", \"fasta\")\n",
    "SeqIO.write(UNKNOWN, \"UNKNOWN_\" + \"viral_all_30up_toSTOP_SEP.fasta\", \"fasta\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2521"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Scanning)+len(IRES)+len(LS)+len(NSP)+len(RS)+len(TR)+len(VPG)+len(UNKNOWN)\n",
    "\n",
    "# Scanning = []\n",
    "\n",
    "\n",
    "# IRES = []\n",
    "\n",
    "# LS = []\n",
    "# NSP = []\n",
    "# RS = []\n",
    "\n",
    "# TR = []\n",
    "\n",
    "# VPG = []\n",
    "\n",
    "# UNKNOWN = []\n",
    "\n",
    "# Scanning[0]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
